src/index.ts

```diff
@@ -161,6 +161,23 @@ export function isOfShape<V extends Dict, T extends Shape<V> = Shape<V>> (shape:
   return fn as GuardWithShape<Unshape<T>>
 }
 
+/**
+ * Create a validator that asserts that passed argument is a tuple of certain elements.
+ * It can validate up to a 3 element tuple.
+ */
+export function isTuple<A>(a: Guard<A>): Guard<[A]>
+export function isTuple<A, B>(a: Guard<A>, b: Guard<B> ): Guard<[A, B]>
+export function isTuple<A, B, C>(a: Guard<A>, b: Guard<B>, c: Guard<C> ): Guard<[A, B, C]>
+export function isTuple<A, B, C>(a: Guard<A>, b?: Guard<B>, c?: Guard<C> ): Guard<[A]> | Guard<[A, B]> | Guard<[A, B, C]> {
+  if (isUndefined(b)) {
+    return ((input: any) => Array.isArray(input) && a(input[0])) as Guard<[A]>
+  } else if (isUndefined(c)) {
+    return ((input: any) => Array.isArray(input) && a(input[0]) && b(input[1])) as Guard<[A, B]>
+  } else {
+    return ((input: any) => Array.isArray(input) && a(input[0]) && b(input[1]) && c(input[2])) as Guard<[A, B, C]>
+  }
```

Comment:
This doesn't scale well. The implementation signature should accept an arbitrary number of arguments, all of type `Guard<any>`. We don't need too much type-safety inside the implementation itself, so just `isTyple (...guards: Guard<Any>): Guard<any>` will be enough. The implementation signature is not visible from the consumer perspective -- when you call `isTuple`, TypeScript will use only the overloads (three in this case) to determine the type. So it's OK for the implementation signature to be more lenient in terms of types.

Then you can iterate through both the given input and the given array of guards. Do an early exit if the input is not an array (you're repeating yourself already with that). Afterwards, check their length; and if that matches, check that each pair matches the type: first guard and first value, second guard and the second value, etc.

This makes implementation work with any number of arguments. The overloads are there "just" to provide super-fancy type-safety for specific cases of 1, 2 or 3 arguments. 

Feel free to ask if the explanation wasn't clear enough.

Reply:
I've applied your suggestions. Please take a look now.